### ISSUE TITLE
Spatialite has multiple issues on 3.12

### ISSUE THREAD

**Issue by Isaacson33** [qgis/QGIS#34932]
After updating to 3.12 I'm experiencing multiple problems with spatialite layers.

1. DB Manager does not show new connections. Making a new connection to a spatialite database doesn't show up in the list of connected databases until after Qgis is exited and restarted.

2. No changes can be made spatialite layers - "could not commit changes" - "unknown cause".

This error occurs on every spatialite database I've tested (8 so far), on three different computers (though all running Manjaro Linux 4.19), it occurs on existing projects, fresh projects and completely new databases.

So - to replicate.

1. Open a fresh Qgis. 
2. Make a new spatialite layer. 
3. Add a feature, save layer.
4. In DB Manager make a new connection to the new database.

At (3) I'm getting the "could not commit changes" error, and at (4) I'm getting no listing of the new database.

**Comment by elpaso** [comment]
Can you please check latest master? I think both issues have been already fixed.

**Comment by uclaros** [comment]
They are fixed indeed in master.
However, you still cannot commit new entries on a table with no geometry and no other fields except the autoincrement pkuid.
```
Could not commit changes to layer geometryless_table

Errors: ERROR: 1 feature(s) not added.
  
  Provider errors:
      SQLite error: unknown cause
    SQL: INSERT INTO "geometryless_table"() VALUES ()
```

On the other hand, while it shouldn't fail to commit such a change, it really makes no sense to add features with no fields and no geometry!

**Comment by Isaacson33** [comment]
> Can you please check latest master? I think both issues have been already fixed.

Only the 3.12.0, 3.10.3 or 3.9.0_master.r54615.7f95991edf-1 packages are available for my Arch-based system. I'm not familiar enough with the build methods to make my own packages directly from source. Unless someone else can test on Arch we'll just have to presume it's fixed for now until the next point release becomes available in the repository.

**Comment by gioman** [comment]
@Isaacson33 both issue were previously reported in other tickets, and fixed... afaik.

**Comment by Isaacson33** [comment]
@gioman 

Can you reference those fixes please. I've tracked down the fix related to the databases not refreshing and applied it. It does indeed fix the first problem. It does not, however, have any impact on the second problem (not being able to commit changes). Which fix addresses this please?

**Comment by elpaso** [comment]
It should be this one: https://github.com/qgis/QGIS/pull/34698

**Comment by Isaacson33** [comment]
> It should be this one: #34698

Ahh, thanks. The last one (which I took to be #34908) I was able to edit 'python/plugins/db_manager/db_plugins/plugin.py' directly with the changes and it fixed the problem. This latest doesn't seem to be that simple. I don't even have a 'src/providers/spatialite/qgsspatialiteprovider.cpp' or a 'tests/src/python/test_provider_spatialite.py' file on my system so I presume they get loaded/altered somehow during the cmake process on linux and I've no idea where to make the changes.

I don't suppose you happen to know where I could find a guide to building Qgis from source for Arch do you? The guides in the Install only cover Debian/Ubuntu.

**Comment by elpaso** [comment]
> I don't suppose you happen to know where I could find a guide to building Qgis from source for Arch do you? The guides in the Install only cover Debian/Ubuntu.

No, sorry.

**Comment by Isaacson33** [comment]
> No, sorry.

No problem, thanks anyway.

Can I ask what might be a naive question (but potentially important, for me), how did this bug get past the testing process? It literally breaks the single most fundamental feature of any GIS - making a polygon. I'm concerned because I use Spatialite for all of my projects and if Spatialite support is so low on the list of priorities that something as simple as making a new polygon doesn't even get checked before a _major_ release it doesn't bode well for more complex or less well-used features.

Is Spatialite being abandoned? If so what should I be moving to in replacement?

**Comment by elpaso** [comment]
@Isaacson33 yours is a fair question and we (developers) have been talking about that several times in the past. The problem with SL data provider is that it has been kind of abandoned during the last few years (we event talked about dropping it completely and use the OGR provider to access SL layers).

Coming to your second question, unfortunately the SL tests on our CI system (Travis) have been disabled for years because they were randomly failing, I'd like to try to re-enable them but since I'm currently 100% in volunteer mode I have no time/resources for that, of course I (and other developers for sure) would be happy to slice some time onto SL provider if someone could fund a few hours of work.

About the editing bug in 3.12, we definitely need more volunteers running manual smoke tests of the upcoming releases to make sure we don't introduce critical bugs (like this one), but 3.12 is not a LTR release and even it it was, we cannot guarantee anything but that the stability in LTR expecially over time during the point release cycle.

**Comment by Isaacson33** [comment]
@elpaso 

Thanks. I'm only a sole-trading consultant with a very small business, but it is heavily reliant on Qgis so it's certainly a possibility for me to consider investing in some time to avoid this in future, it would depend how much of an investment would be needed.

I will talk to some of my bigger clients - some rely on Spatialite-based solutions I've made for them, they might be willing to fund something, but I'd need to persuade them it's worth it in the long-run. If Spatialite isn't going to last as a format then they'd probably rather focus on funding replacements (like improving the SQL capabilities of Geopackages, for example).

The problem is if the SL data provider has been abandoned and there's little interest in SL testing, then is it a worthwhile route to take?

I've thought Spatialite support has been lacking in Qgis for some time (in that Spatialite bugs were often slow to get fixed) and I suppose what you've said about the data provider and testing could explain a lot of that.

The trouble is, I don't personally know of any other solution to self-hosted geospatial databases which allow SQL queries. I have PostGIS on my home system, but it's way overkill for a self-hosted database, Geopackages aren't there yet in terms of the SQL queries they support, so what other options are people using such that they can abandon Spatialite?

As to manual testing - I moved to Manjaro Linux last year and I haven't yet figured out quite how I can download and install nightly releases. I think I have to learn how to make packages out of the git repository but that's been on the to-do list for a while and other work keeps getting in the way. As soon as I manage to get over that hurdle, I'd be happy to get involved running some tests.

**Comment by elpaso** [comment]
@Isaacson33 I would say it is abandoned **now**: see https://github.com/qgis/QGIS/commits/master/src/providers/spatialite and we are not dropping it anytime soon.

But it has certainly been lagging behind in term of attention from most of us.

I believe the highest priority would be to have CI tests back online, that would certainly reduce the probability of incidents like this one.

I think the best candidate for a replacement is GPKG, but I'm not sure it has quite the same similar SQL capabilities.

**Comment by Isaacson33** [comment]
> I believe the highest priority would be to have CI tests back online, that would certainly reduce the probability of incidents like this one.

OK, what size project do you think it would be to get that working again? How much funding would be needed?

**Comment by elpaso** [comment]
Hard to tell, best case a few hours, but if we get random failures again it's very hard and time consuming to debug them.

**Comment by Isaacson33** [comment]
I'm sure Qgis used to have a system for funding specific developments, but I can't seem to find anything on it now I'm looking. Have they abandoned the practice?

**Comment by elpaso** [comment]
AFAIK there are never been earmarked fundings for specific bugs, you can still donate for generic bugfixing, but that would not mean that the funds go to a specific part of the code.

What other customers usually do in such cases is to contract a core developer to fix a specific issue.

Please don't get me wrong, I'm not saying that myself or any other developer won't eventually dedicate some time on this task, just that there is no guarantee that it will be happen in a certain time: it may happen tonight or in ten years, nobody knows.

In case you missed it, @nyalldawson wrote a comprehensive post about the topic a few years back: https://nyalldawson.net/2016/08/how-to-effectively-get-things-changed-in-qgis/

**Comment by jef-n** [comment]
> I think the best candidate for a replacement is GPKG, but I'm not sure it has quite the same similar SQL capabilities.

Doesn't GDAL's gpkg support also depend on spatialite?

**Comment by elpaso** [comment]
@jef-n no: QGIS data provider implementation for GPKG is all inside the OGR provider.

**Comment by jef-n** [comment]
> @jef-n no: QGIS data provider implementation for GPKG is all inside the OGR provider.

GDAL's gpkg support != our gpkg support.

**Comment by rouault** [comment]
@jef-n GDAL GPKG driver *optionally* uses spatialite to enable it on GPKG databases it opens. There is no really use of it in the code of the driver. It's just that users may use spatialite functions with ExecuteSQL()

**Comment by uclaros** [comment]
Could we use the OGR provider for spatialite too?

@Isaacson33 what are the limitations of GPKG in your use cases?

**Comment by elpaso** [comment]
@uclaros you can use OGR to open a SL file right now, just open a spatialite file from the QGIS file browser.

**Comment by Isaacson33** [comment]
@elpaso 
> AFAIK there are never been earmarked fundings for specific bugs, you can still donate for generic bugfixing, but that would not mean that the funds go to a specific part of the code.

I must have misremembered, that would explain why I can't find anything about it.

> What other customers usually do in such cases is to contract a core developer to fix a specific issue.

That's what I'll have to do then, thanks for pointing me in the right direction.

@uclaros 
> what are the limitations of GPKG in your use cases?

It's been a while since I checked, but off the top of my head - grouping functions (sum, max, min...), joining tables in queries, certain geometry functions (st_buffer, st_within...). Those were the ones I recall blocking any further progress, there might be others.

**Comment by gioman** [comment]
> It's been a while since I checked, but off the top of my head - grouping functions (sum, max, min...), joining tables in queries, certain geometry functions (st_buffer, st_within...). Those were the ones I recall blocking any further progress, there might be others.

I agree that using spatial SQL functions is (was?) cumbersome with GPKGs in DB Manager, but I also not checked if something changed for the better in the latest releases.

**Comment by elpaso** [comment]
@gioman IIRC something was done to bring "spatialite" powers to GPKG recently. CC  @roya0045 ?

**Comment by roya0045** [comment]
@elpaso Yes this was the case. Though the spatialite functions were enabled upstream in OGR by Evan shortly before or after I made the PR, so I only added the support in the DB Manager afterall. Meaning that geospatial functions should now work in geopackages.

**Comment by Isaacson33** [comment]
I've just downgraded to the LTR 3.10.3 https://qgis.org/downloads/qgis-3.10.3.tar.bz2, and the issue seems to be still there. Maybe I've misunderstood, but aren't fixes for breaking bugs supposed to be backported to the LTR?

If so, then this issue is not fixed (for me, anyway).

- Unless... the LTR packaged for Arch is not the latest version with the backport? If so, it should be updated on the qgis/downloads.

**Comment by elpaso** [comment]
@Isaacson33 it was backported, you need to wait for the next point release.

**Comment by Isaacson33** [comment]
@elpaso 

Sorry, I'm not following you. If it was backported, then shouldn't 3.10.3 work? Yet you say I need to wait for the next point release (not an option unfortunately - work commitments - I have to keep trying to get _something_ working).

**Comment by elpaso** [comment]
It will be in 3.10.4, you can probably use the nightly builds now.

**Comment by Isaacson33** [comment]
> It will be in 3.10.4, you can probably use the nightly builds now.

No, the nightly builds are running into multiple compile errors on Linux (all seem to be around the ibqgis_gui.so libraries), so they're not a solution right now.

3.10.3 is still listed and packaged as the LTR, so, am I correct in thinking that it _has_ been backported, but that the  LTR git repository has not yet been packaged for release to any installers?

**Comment by elpaso** [comment]
> 3.10.3 is still listed and packaged as the LTR, so, am I correct in thinking that it _has_ been backported, but that the LTR git repository has not yet been packaged for release to any installers?

exactly.

**Comment by Isaacson33** [comment]
> exactly.

OK, thanks. I suppose I'll have to keep on at trying to track down the compile errors from cloning the latest git then. It could well be weeks/months before the next point release actually makes it the Arch repos.

**Comment by elpaso** [comment]
you can try to run QGIS desktop with docker too, see:  https://github.com/qgis/QGIS/tree/master/.docker

**Comment by Isaacson33** [comment]
> you can try to run QGIS desktop with docker too, see: https://github.com/qgis/QGIS/tree/master/.docker

Thanks. I've not tried anything like that before, but at this point anything's worth a go!

**Comment by uclaros** [comment]
@Isaacson33 I think you've misunderstood backporting.
For clarification here's a sample timeline:
- 3.12.0 is the newest version, 3.10.3 is the current LTR
- Bug is fixed in master.
- Bugfix will appear on 3.12.1
- If backported, it will also appear on 3.10.4
- Arch repos package 3.12.1 and 3.10.4 LTR so you can update.

**Comment by gioman** [comment]
> 3.12.0 is the newest version, 3.10.3 is the current LTR

isn't just more clear for the users to state that 3.10 is the LTR? and that this version will get a constant flow of fixes/backports that translates into point releases.

**Comment by uclaros** [comment]
> isn't just more clear for the users to state that 3.10 is the LTR? and that this version will get a constant flow of fixes/backports that translates into point releases.
Indeed!

My point was that point releases do not change. 3.10.3 will always contain the bug. You'll have to wait for next point release, build from source or use a pre built nightly.

**Comment by Isaacson33** [comment]
@uclaros , @gioman 

Thanks. That's all very clear now, but I can't wait for the point releases to get packaged in order to deal with a bug so big as to render Spatialite completely unusable. I've never been in this situation before despite over a decade of happy Qgis use. Bugs in the latest versions have always been annoying but tolerable and I've just waited for the next version to arrive. This is the first time a bug has been allowed through to a release which actually breaks the application completely for Spatialite.

As such, I'm in new territory, and following what @elpaso has said about the absence of Spatialite testing, it sounds like I'm going to be here for some time, with new releases potentially being completely unuseable with Spatialite and, from the sound of it, the ltr isn't even safe because a backport could be commited to it which breaks Spatialite (as an unintended consequence, of course) and no-one would know before release.

Basically, if I'm going to continue to use Spatialite (which I have to for work), and Qgis is going to continue without any testing at all to check if Saptialite works on it's releases, then the only way forward I see is for me to try and run the master (or ltr latest git) directly.

So that's what I'm currently trying to learn how to do. As you can imagine, it's quite a steep learning curve, so I appreciate you setting me right where I've misunderstood some aspect.

**Comment by uclaros** [comment]
@Isaacson33 If you have trouble building from source or using the docker image, maybe you could just downgrade to 3.10.2 and not upgrade until 3.10.4 comes out in about a week (Time until next pointrelease 2020-03-20 12:00:00 UTC 7d 3h 7m).

**Comment by haubourg** [comment]
@Isaacson33 I had to use docker extensively here because of the recent proj/gdal upgrades that makes it very hard to compile on ubuntu 18. 
My docker recipes are here :+1: 

for master:
https://github.com/haubourg/docker-qgis-desktop/tree/3.x_master

and there is a branch for each version. 

Don't hesitate to ask for help

**Comment by Isaacson33** [comment]
@uclaros 
The step in the process that I'm currently stuck on is making a proper PKGBUILD file for the tarball to install it on my Arch-based system, so I'd be just as stuck doing that for the 3.10.2 tar as I would using the master git as a source, but thanks for the suggestion anyway.

@haubourg 
I really appreciate the offer of help. I have to say that I've no experience at all with dockers - I only even found out what they were yesterday (and I'm not even clear on that). Are they fairly easy to manage? The Arch wiki on them, as ever, makes it look extremely complicated.

**Comment by uclaros** [comment]
@Isaacson33 I have no experience with Arch, but I think when you install packages with *pacman* you can choose to install previous versions from the arch linux archive https://wiki.archlinux.org/index.php/Arch_Linux_Archive

**Comment by Isaacson33** [comment]
> @Isaacson33 I have no experience with Arch, but I think when you install packages with _pacman_ you can choose to install previous versions from the arch linux archive https://wiki.archlinux.org/index.php/Arch_Linux_Archive

Thank you for for this. I did actually come across the ALA sometime back but had completely forgotten about it. I was so focussed on getting a newer version (or a newer LTR) I didn't think of going even further backwards. Plus, downgrades have generally failed for me in the past (shared library problems), but this one worked flawlessly. That's solved two thirds of my problems.

I still have some projects which I need to work on using the 3.12.x version, but I'm going to try the docker for that - maybe next week though!

**Comment by uclaros** [comment]
Maybe this could be of help for the other third of your problem: In my simple test with spatialite in 3.12.0 I saw that I could properly commit if I completed by hand the autoincrement field pkuid.

**Comment by Isaacson33** [comment]
> Maybe this could be of help for the other third of your problem: In my simple test with spatialite in 3.12.0 I saw that I could properly commit if I completed by hand the autoincrement field pkuid.

That would be great if it worked. A colleague of mine is having the same trouble with 3.12 on his laptop and I don't think he's got round to downgrading yet. I might try it out when I can. It would be typical of my computing experience in general if this whole episode could have been solved in one simple adjustment!
